storage/remote/queue_manager.go

```diff
@@ -1078,23 +1071,21 @@ func (s *shards) runShard(ctx context.Context, shardID int, queue chan interface
 			// stop reading from the queue. This makes it safe to reference pendingSamples by index.
 			switch d := sample.(type) {
 			case writeSample:
-				sampleBuffer[nPendingSamples][0] = d.sample
 				pendingData[nPending].Labels = labelsToLabelsProto(d.seriesLabels, pendingData[nPending].Labels)
-				pendingData[nPending].Samples = sampleBuffer[nPendingSamples]
+				pendingData[nPending].Samples = []prompb.Sample{d.sample}
```

Comment:
By creating a new slice of samples I fear that we are reducing some of the allocation optimizations made in https://github.com/prometheus/prometheus/pull/5849. I wonder if we could still pre-allocate samples with len=0 and cap=1, then either set the value or truncate the slice as we use them?

Here is a benchmark you can use to see these concerns:
```
name               old time/op    new time/op    delta
SampleDelivery-16     15.5s ± 2%     16.9s ± 1%   +8.72%  (p=0.001 n=7+6)

name               old alloc/op   new alloc/op   delta
SampleDelivery-16    20.5GB ± 0%    21.7GB ± 0%   +5.85%  (p=0.001 n=7+7)

name               old allocs/op  new allocs/op  delta
SampleDelivery-16      176M ± 0%      201M ± 0%  +14.20%  (p=0.001 n=7+7)
```

Reply:
I think what you're saying is that when we allocate `pendingData` we should also allocate len=0 series and exemplar slices to pre-populate the fields. The concern there was we'd over-allocate exemplars, but I think that if an allocation now happens for every single new sample, the initial penalty is well worth it.

Reply:
Yep, if someone has send exemplars turned on chances are they have some exemplars so eventually all of those would be allocated anyway, and it is much better than creating garbage that needs to be cleaned up for every sample.

Reply:
Actually, after trying to implement this I don't think it will work. Before when there were two separate slices they would only get sent up the index they had been filled, eliminated the possibility of a sample from the last cycle being sent again. But, because now each slice can hold _either_ a series or exemplar, the sample from the last cycle won't get wiped out if they aren't of the same type as the new sample. The only way no samples get re-sent is if the slice is filled in the exact same order every single time, which isn't guaranteed.

Reply:
Hmm, yeah this may add a fair bit more complexity. My thought is something like:
```
pendingData[nPending].Samples = pendingData[nPending].Samples[:0]
pendingData[nPending].Exemplars = pendingData[nPending].Exemplars[:0]
switch d := sample.(type) {
case writeSample:
    pendingData[nPending].Samples = append(pendingData[nPending].Samples, d.sample)
case writeExemplar:
    pendingData[nPending].Exemplars = append(pendingData[nPending].Exemplars, d.exemplar)
}
```
Do you think that would work? You might be able to use some ifs to avoid the truncations of the slices, but I don't know if it would help with performance at all.

Reply:
I'm not sure, let's give it a try. Just pushed the change, tests pass. I'm not sure what benchmark you were using or else I would run it myself.

Reply:
I have been running:
```
go test -bench=BenchmarkSampleDelivery -benchmem -run=djfkasldf ./storage/remote
```
for the benchmark. Here are the results for before this change, without preallocation, and with preallocation:
```
benchstat before.txt no_preallocation.txt preallocation.txt
name \ time/op     before.txt   no_preallocation.txt  preallocation.txt
SampleDelivery-16   15.5s ± 2%            16.9s ± 1%         15.1s ± 1%

name \ alloc/op    before.txt   no_preallocation.txt  preallocation.txt
SampleDelivery-16  20.5GB ± 0%           21.7GB ± 0%        20.5GB ± 0%

name \ allocs/op   before.txt   no_preallocation.txt  preallocation.txt
SampleDelivery-16    176M ± 0%             201M ± 0%          176M ± 0%
```

So looks like we are back to the same amount of allocations now (yay!) and might even be a couple percent faster compared to before this PR.